proxy/src/auth/backend/jwt.rs

```diff
@@ -0,0 +1,549 @@
+use std::{future::Future, sync::Arc, time::Duration};
+
+use anyhow::{bail, ensure, Context};
+use arc_swap::ArcSwapOption;
+use bytes::Bytes;
+use dashmap::DashMap;
+use jose_jwk::crypto::KeyInfo;
+use serde::de::DeserializeOwned;
+use signature::Verifier;
+use tokio::time::Instant;
+
+use crate::intern::EndpointIdInt;
+
+#[derive(Default)]
+pub struct JWKCache {
+    client: reqwest::Client,
+
+    map: DashMap<EndpointIdInt, Arc<JWKCacheEntryLock>>,
+}
+
+pub struct JWKCacheEntryLock {
+    cached: ArcSwapOption<JWKCacheEntry>,
+    lookup: tokio::sync::Semaphore,
+}
+
+impl Default for JWKCacheEntryLock {
+    fn default() -> Self {
+        JWKCacheEntryLock {
+            cached: ArcSwapOption::empty(),
+            lookup: tokio::sync::Semaphore::new(1),
```

Comment:
We could use an atomic ID system and panic if the IDs are mismatched. I suppose even the pointer address could double as the ID. I think this might be excessive though